proposals/0000-gadt-syntax.rst

```diff
@@ -0,0 +1,320 @@
+
+* It is too permissive with regards to parentheses, causing implementation
+  issues (see `#19192
+  <https://gitlab.haskell.org/ghc/ghc/-/issues/19192#note_329172>`_).
+
+In this proposal we aim to give an alternative specification that would fix the
+aforementioned issues.
+
+Let us first point out that GADT constructor syntax pursues two contradictory
+goals:
+
+1. On the one hand, GADT constructor syntax tries its best to mimic type
+   syntax::
+
+     data T x where
+       MkT :: forall a. Eq a => Maybe a -> T 0
+
+     ghci> :t +v MkT
+     MkT :: forall a. Eq a => Maybe a -> T 0
+
+   Notice how the type signature reported by ``:t`` is identical to the one in
+   the declaration.
+
+2. On the other hand, GADT constructor syntax allows the user to specify
+   additional information about the fields, such as their names (using record
+   syntax ``{ fld :: a }``) and strictness (``{-# UNPACK #-}``, ``!a``,
+   ``~a``). This information is not reflected in the constructor's type.
+
+It is impossible to define a constructor by its type signature and at the same
+time include information that is not part of its type signature. This inherent
+contradiction means that any solution will be a compromise.
+
+Perhaps if we were to drop one of these goals, we could come up with something
+glorious, but in practice we must take backwards-compatibility and ease of
+implementation into account. With that in mind, let us establish the following
+principles:
+
+1. A GADT constructor signature is either in *prefix style* or *record style*.
+   The two styles are completely distinct and we do not seek to unify their syntax.
+
+2. A record-style GADT constructor signature has the following parts in a
+   fixed order:
+
+   #. (optional) ``forall tvs.`` to bind type variables;
+   #. (optional) ``ctx =>`` to introduce constraints;
+   #. ``{ fld1 :: a, fld2 :: b, ... }`` to describe constructor fields, using
+      the same syntax as classic Haskell98-style record type declarations, such
+      as ``data R = MkR { fld1 :: a, fld2 :: b, ... }``;
+   #. ``->`` as a separator, which is special syntax rather than
+      the ``(->)`` type constructor;
+   #. the result type ``T``;
+
+3. A prefix-style GADT constructor signature has zero or more of the following
+   parts in a free order:
+
+   #. type variable telescope ``forall tvs.`` or ``forall tvs ->``;
+   #. constraint context ``ctx =>``;
+   #. parameter ``a ->``, ``a %p ->``, or ``a ⊸``, possibly with a strictness
+      annotation;
+
+   It then ends with the result type ``T``.
+
+
+Proposed Change Specification
+-----------------------------
+
+Changes from status quo
+~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Drop support for parentheses around the tail of a prefix-style constructor
+   signature (fix ``#19192``).
+2. Permit nested foralls and contexts (a free order of quantifiers) in
+   prefix-style constructor signatures (fix ``#18389``).
```

Comment:
Very cheap nitpicking on my part. But it would be quite helpful if you added links to the issues here (and below).